### PR TITLE
Merge and clean up SparseQArray with the rest of QArray

### PR DESCRIPTION
**Pull request by gautierronan**
I suggest reading commit by commit: it's a bunch of minor stuff meant to uniformize the implementation with the rest of QArray.

### CODE REVIEW

**Review by pierreguilmin — COMMENTED** [review]
Apart from the discussion on wheter we should revert 192a341 (which would modify this PR), happy to merge.

**Comment by gautierronan** [comment]
Added a few more commits, all rather small. Ideally we have tests passing for `tests/qarrays/test_sparse_qarray.py` which is still not the case. 

> Apart from the discussion on wheter we should revert [192a341](https://github.com/dynamiqs/dynamiqs/commit/192a34147a23b1580cc6b366f3a0d9ff451dc1a5) (which would modify this PR), happy to merge.

We can revert in a separate PR, but we should discuss the sparse API before that IMO.

File: dynamiqs/time_array.py [hunk 1]
```diff
@@ -78,7 +78,6 @@ def pwc(times: ArrayLike, values: ArrayLike, array: ArrayLike) -> PWCTimeArray:
         )
 
     # array
-    array = jnp.asarray(array, dtype=cdtype())
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:81]: We need this back I think.

File: dynamiqs/time_array.py [hunk 2]
```diff
@@ -109,7 +108,6 @@ def modulated(f: callable[[float, ...], Array], array: ArrayLike) -> ModulatedTi
...
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:112]: Also here.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]

File: dynamiqs/time_array.py (hunk 1 above)
Comment (gautierronan) [reply dynamiqs/time_array.py]: Good catch

**Pull request merged**